In [1]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", 150)
pd.set_option("display.max_rows", 100)

RAW_PATH = "../data/raw/LoanStats_2018Q2.csv"

df = pd.read_csv(RAW_PATH, skiprows=1, low_memory=False)

df.shape

(130774, 145)

## Initial Data Inspection

In [2]:
df.head()

,id,member_id,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,emp_title,emp_length,home_ownership,annual_inc,verification_status,issue_d,loan_status,pymnt_plan,url,desc,purpose,title,zip_code,addr_state,dti,delinq_2yrs,earliest_cr_line,inq_last_6mths,mths_since_last_delinq,mths_since_last_record,open_acc,pub_rec,revol_bal,revol_util,total_acc,initial_list_status,out_prncp,out_prncp_inv,total_pymnt,total_pymnt_inv,total_rec_prncp,total_rec_int,total_rec_late_fee,recoveries,collection_recovery_fee,last_pymnt_d,last_pymnt_amnt,next_pymnt_d,last_credit_pull_d,collections_12_mths_ex_med,mths_since_last_major_derog,policy_code,application_type,annual_inc_joint,dti_joint,verification_status_joint,acc_now_delinq,tot_coll_amt,tot_cur_bal,open_acc_6m,open_act_il,open_il_12m,open_il_24m,mths_since_rcnt_il,total_bal_il,il_util,open_rv_12m,open_rv_24m,max_bal_bc,all_util,total_rev_hi_lim,inq_fi,total_cu_tl,inq_last_12m,acc_open_past_24mths,avg_cur_bal,bc_open_to_buy,bc_util,chargeoff_within_12_mths,delinq_amnt,mo_sin_old_il_acct,mo_sin_old_rev_tl_op,mo_sin_rcnt_rev_tl_op,mo_sin_rcnt_tl,mort_acc,mths_since_recent_bc,mths_since_recent_bc_dlq,mths_since_recent_inq,mths_since_recent_revol_delinq,num_accts_ever_120_pd,num_actv_bc_tl,num_actv_rev_tl,num_bc_sats,num_bc_tl,num_il_tl,num_op_rev_tl,num_rev_accts,num_rev_tl_bal_gt_0,num_sats,num_tl_120dpd_2m,num_tl_30dpd,num_tl_90g_dpd_24m,num_tl_op_past_12m,pct_tl_nvr_dlq,percent_bc_gt_75,pub_rec_bankruptcies,tax_liens,tot_hi_cred_lim,total_bal_ex_mort,total_bc_limit,total_il_high_credit_limit,revol_bal_joint,sec_app_earliest_cr_line,sec_app_inq_last_6mths,sec_app_mort_acc,sec_app_open_acc,sec_app_revol_util,sec_app_open_act_il,sec_app_num_rev_accts,sec_app_chargeoff_within_12_mths,sec_app_collections_12_mths_ex_med,sec_app_mths_since_last_major_derog,hardship_flag,hardship_type,hardship_reason,hardship_status,deferral_term,hardship_amount,hardship_start_date,hardship_end_date,payment_plan_start_date,hardship_length,hardship_dpd,hardship_loan_status,orig_projected_additional_accrued_interest,hardship_payoff_balance_amount,hardship_last_payment_amount,disbursement_method,debt_settlement_flag,debt_settlement_flag_date,settlement_status,settlement_date,settlement_amount,settlement_percentage,settlement_term
0,NaN,NaN,12000.0,12000.0,12000.0,60 months,20.39%,320.54,D,D4,Garage Man,10+ years,RENT,42000.0,Verified,Jun-2018,Current,n,NaN,NaN,credit_card,Credit card refinancing,440xx,OH,14.46,0.0,Nov-2007,1.0,NaN,NaN,4.0,0.0,8090.0,64.7%,9.0,w,11269.75,11269.75,1896.05,1896.05,730.25,1165.80,0.0,0.0,0.0,Jan-2019,320.54,Feb-2019,Jan-2019,0.0,NaN,1.0,Individual,NaN,NaN,NaN,0.0,0.0,11968.0,1.0,1.0,0.0,1.0,24.0,3878.0,33.0,3.0,3.0,5445.0,50.0,12500.0,1.0,0.0,1.0,4.0,2992.0,4410.0,64.7,0.0,0.0,125.0,127.0,3.0,3.0,0.0,3.0,NaN,3.0,NaN,0.0,3.0,3.0,3.0,4.0,4.0,3.0,5.0,3.0,4.0,0.0,0.0,0.0,3.0,100.0,66.7,0.0,0.0,24138.0,11968.0,12500.0,11638.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
1,NaN,NaN,29150.0,29150.0,29150.0,60 months,20.39%,778.64,D,D4,Police Officer,10+ years,MORTGAGE,51000.0,Verified,Jun-2018,Current,n,NaN,NaN,home_improvement,Home improvement,807xx,CO,20.82,0.0,Feb-2003,1.0,49.0,NaN,11.0,0.0,12760.0,40.4%,34.0,w,27376.13,27376.13,4605.80,4605.80,1773.87,2831.93,0.0,0.0,0.0,Jan-2019,778.64,Feb-2019,Jan-2019,0.0,NaN,1.0,Individual,NaN,NaN,NaN,0.0,0.0,177088.0,2.0,1.0,0.0,3.0,13.0,18790.0,83.0,3.0,3.0,3052.0,58.0,31600.0,0.0,0.0,2.0,7.0,16099.0,16854.0,39.6,0.0,0.0,184.0,176.0,1.0,1.0,4.0,1.0,NaN,3.0,NaN,0.0,7.0,8.0,8.0,12.0,15.0,9.0,15.0,8.0,11.0,0.0,0.0,0.0,4.0,97.1,37.5,0.0,0.0,201814.0,31550.0,27900.0,22714.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,8000.0,8000.0,8000.0,36 months,13.06%,269.79,C,C1,Manager,10+ years,MORTGAGE,73000.0,Source Verified,Jun-2018,Current,n,NaN,NaN,other,O

In [3]:
df.tail()

,id,member_id,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,emp_title,emp_length,home_ownership,annual_inc,verification_status,issue_d,loan_status,pymnt_plan,url,desc,purpose,title,zip_code,addr_state,dti,delinq_2yrs,earliest_cr_line,inq_last_6mths,mths_since_last_delinq,mths_since_last_record,open_acc,pub_rec,revol_bal,revol_util,total_acc,initial_list_status,out_prncp,out_prncp_inv,total_pymnt,total_pymnt_inv,total_rec_prncp,total_rec_int,total_rec_late_fee,recoveries,collection_recovery_fee,last_pymnt_d,last_pymnt_amnt,next_pymnt_d,last_credit_pull_d,collections_12_mths_ex_med,mths_since_last_major_derog,policy_code,application_type,annual_inc_joint,dti_joint,verification_status_joint,acc_now_delinq,tot_coll_amt,tot_cur_bal,open_acc_6m,open_act_il,open_il_12m,open_il_24m,mths_since_rcnt_il,total_bal_il,il_util,open_rv_12m,open_rv_24m,max_bal_bc,all_util,total_rev_hi_lim,inq_fi,total_cu_tl,inq_last_12m,acc_open_past_24mths,avg_cur_bal,bc_open_to_buy,bc_util,chargeoff_within_12_mths,delinq_amnt,mo_sin_old_il_acct,mo_sin_old_rev_tl_op,mo_sin_rcnt_rev_tl_op,mo_sin_rcnt_tl,mort_acc,mths_since_recent_bc,mths_since_recent_bc_dlq,mths_since_recent_inq,mths_since_recent_revol_delinq,num_accts_ever_120_pd,num_actv_bc_tl,num_actv_rev_tl,num_bc_sats,num_bc_tl,num_il_tl,num_op_rev_tl,num_rev_accts,num_rev_tl_bal_gt_0,num_sats,num_tl_120dpd_2m,num_tl_30dpd,num_tl_90g_dpd_24m,num_tl_op_past_12m,pct_tl_nvr_dlq,percent_bc_gt_75,pub_rec_bankruptcies,tax_liens,tot_hi_cred_lim,total_bal_ex_mort,total_bc_limit,total_il_high_credit_limit,revol_bal_joint,sec_app_earliest_cr_line,sec_app_inq_last_6mths,sec_app_mort_acc,sec_app_open_acc,sec_app_revol_util,sec_app_open_act_il,sec_app_num_rev_accts,sec_app_chargeoff_within_12_mths,sec_app_collections_12_mths_ex_med,sec_app_mths_since_last_major_derog,hardship_flag,hardship_type,hardship_reason,hardship_status,deferral_term,hardship_amount,hardship_start_date,hardship_end_date,payment_plan_start_date,hardship_length,hardship_dpd,hardship_loan_status,orig_projected_additional_accrued_interest,hardship_payoff_balance_amount,hardship_last_payment_amount,disbursement_method,debt_settlement_flag,debt_settlement_flag_date,settlement_status,settlement_date,settlement_amount,settlement_percentage,settlement_term
130769,NaN,NaN,14000.0,14000.0,13975.0,36 months,16.01%,492.27,C,C5,client analyst,10+ years,MORTGAGE,52874.0,Not Verified,Apr-2018,Current,n,NaN,NaN,debt_consolidation,Debt consolidation,912xx,CA,22.36,0.0,Sep-1995,1.0,NaN,87.0,9.0,1.0,25426.0,85%,20.0,f,11099.23,11079.41,4417.98,4410.09,2900.77,1517.21,0.0,0.0,0.0,Jan-2019,492.27,Feb-2019,Jan-2019,0.0,NaN,1.0,Individual,NaN,NaN,NaN,0.0,0.0,32056.0,2.0,1.0,1.0,1.0,6.0,6630.0,88.0,1.0,2.0,8552.0,85.0,29900.0,1.0,1.0,1.0,3.0,3561.0,5520.0,91.0,0.0,0.0,6.0,270.0,4.0,4.0,0.0,15.0,NaN,6.0,NaN,0.0,5.0,8.0,10.0,10.0,1.0,8.0,19.0,7.0,9.0,NaN,0.0,0.0,2.0,100.0,80.0,1.0,0.0,37400.0,32056.0,27600.0,7500.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
130770,NaN,NaN,7500.0,7500.0,7500.0,36 months,10.90%,245.19,B,B4,Residence Counselor,10+ years,RENT,126000.0,Not Verified,Apr-2018,Current,n,NaN,NaN,other,Other,125xx,NY,6.33,0.0,Dec-1986,1.0,NaN,NaN,16.0,0.0,11122.0,28.3%,31.0,f,5847.27,5847.27,2197.63,2197.63,1652.73,544.90,0.0,0.0,0.0,Jan-2019,245.19,Feb-2019,Jan-2019,0.0,NaN,1.0,Individual,NaN,NaN,NaN,0.0,0.0,132066.0,1.0,0.0,0.0,0.0,68.0,0.0,NaN,5.0,8.0,5564.0,28.0,39300.0,1.0,0.0,2.0,11.0,8254.0,23906.0,31.1,0.0,0.0,146.0,375.0,1.0,1.0,3.0,1.0,NaN,1.0,NaN,0.0,9.0,11.0,11.0,16.0,4.0,15.0,24.0,11.0,16.0,0.0,0.0,0.0,5.0,100.0,0.0,0.0,0.0,162600.0,11122.0,34700.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
130771,NaN,NaN,35000.0,35000.0,35000.0,36 months,6.07%,1065.88,A,A2,Founder,10+ years,RENT,110000.0,Source Verified,Apr-2018,Current,n,NaN,NaN,debt_c

#### Observation

The CSV loads into 130,774 rows and 145 columns after skipping the first note row. The last two rows are not real loan records; they are LendingClub funding summary rows. These should be removed before modeling or final analysis, which leaves 130,772 actual loan records.

In [4]:
df.info(verbose=True, show_counts=True)

<class 'pandas.DataFrame'>
RangeIndex: 130774 entries, 0 to 130773
Data columns (total 145 columns):
 #    Column                                      Non-Null Count   Dtype  
---   ------                                      --------------   -----  
 0    id                                          2 non-null       str    
 1    member_id                                   0 non-null       float64
 2    loan_amnt                                   130772 non-null  float64
 3    funded_amnt                                 130772 non-null  float64
 4    funded_amnt_inv                             130772 non-null  float64
 5    term                                        130772 non-null  str    
 6    int_rate                                    130772 non-null  str    
 7    installment                                 130772 non-null  float64
 8    grade                                       130772 non-null  str    
 9    sub_grade                                   130772 non-null  str    

In [5]:
column_summary = pd.DataFrame({
  "dtype":df.dtypes,
  "non_null": df.notna().sum(),
  "missing": df.isna().sum(),
  "missing_pct": df.isna().mean() * 100
})

#column_summary.sort_values("missing_pct", ascending=False)
column_summary.sort_index()

,dtype,non_null,missing,missing_pct
acc_now_delinq,float64,130772,2,0.001529
acc_open_past_24mths,float64,130772,2,0.001529
addr_state,str,130772,2,0.001529
all_util,float64,130737,37,0.028293
annual_inc,float64,130772,2,0.001529
...,...,...,...,...
total_rev_hi_lim,float64,130772,2,0.001529
url,float64,0,130774,100.000000
verification_status,str,130772,2,0.001529
verification_status_joint,str,17417,113357,86.681603


#### Observation

Several columns are fully or almost fully missing. `member_id`, `url`, and `desc` contain no usable values, while `id` only appears in the two footer summary rows. Some hardship and settlement columns are also extremely sparse, so they are likely candidates for dropping.

In [6]:
df.duplicated().sum()

np.int64(0)

In [7]:
df["loan_status"].value_counts(dropna=False)

loan_status
Current               113004
Fully Paid             11531
Late (31-120 days)      2418
Charged Off             1838
In Grace Period         1353
Late (16-30 days)        627
NaN                        2
Default                    1
Name: count, dtype: int64

#### Observation

`loan_status` is heavily dominated by `Current` loans. This matters because many 2018Q2 loans had not reached a final outcome by the snapshot date, so `Current` should not be treated the same as a fully resolved good or bad loan outcome without thinking about the modeling target.

In [8]:
df.describe(percentiles=[0.01, 0.05, 0.25, 0.5, 0.75, 0.95, 0.99]).round(2)

,member_id,loan_amnt,funded_amnt,funded_amnt_inv,installment,annual_inc,url,desc,dti,delinq_2yrs,inq_last_6mths,mths_since_last_delinq,mths_since_last_record,open_acc,pub_rec,revol_bal,total_acc,out_prncp,out_prncp_inv,total_pymnt,total_pymnt_inv,total_rec_prncp,total_rec_int,total_rec_late_fee,recoveries,collection_recovery_fee,last_pymnt_amnt,collections_12_mths_ex_med,mths_since_last_major_derog,policy_code,annual_inc_joint,dti_joint,acc_now_delinq,tot_coll_amt,tot_cur_bal,open_acc_6m,open_act_il,open_il_12m,open_il_24m,mths_since_rcnt_il,total_bal_il,il_util,open_rv_12m,open_rv_24m,max_bal_bc,all_util,total_rev_hi_lim,inq_fi,total_cu_tl,inq_last_12m,acc_open_past_24mths,avg_cur_bal,bc_open_to_buy,bc_util,chargeoff_within_12_mths,delinq_amnt,mo_sin_old_il_acct,mo_sin_old_rev_tl_op,mo_sin_rcnt_rev_tl_op,mo_sin_rcnt_tl,mort_acc,mths_since_recent_bc,mths_since_recent_bc_dlq,mths_since_recent_inq,mths_since_recent_revol_delinq,num_accts_ever_120_pd,num_actv_bc_tl,num_actv_rev_tl,num_bc_sats,num_bc_tl,num_il_tl,num_op_rev_tl,num_rev_accts,num_rev_tl_bal_gt_0,num_sats,num_tl_120dpd_2m,num_tl_30dpd,num_tl_90g_dpd_24m,num_tl_op_past_12m,pct_tl_nvr_dlq,percent_bc_gt_75,pub_rec_bankruptcies,tax_liens,tot_hi_cred_lim,total_bal_ex_mort,total_bc_limit,total_il_high_credit_limit,revol_bal_joint,sec_app_inq_last_6mths,sec_app_mort_acc,sec_app_open_acc,sec_app_revol_util,sec_app_open_act_il,sec_app_num_rev_accts,sec_app_chargeoff_within_12_mths,sec_app_collections_12_mths_ex_med,sec_app_mths_since_last_major_derog,deferral_term,hardship_amount,hardship_length,hardship_dpd,orig_projected_additional_accrued_interest,hardship_payoff_balance_amount,hardship_last_payment_amount,settlement_amount,settlement_percentage,settlement_term
count,0.0,130772.00,130772.00,130772.00,130772.00,130772.00,0.0,0.0,130446.00,130772.00,130772.00,58400.00,16558.00,130772.00,130772.00,130772.00,130772.00,130772.00,130772.00,130772.00,130772.00,130772.00,130772.00,130772.00,130772.00,130772.00,130772.00,130772.00,31062.00,130772.0,18087.00,18087.00,130772.0,130772.00,130772.00,130772.00,130772.00,130772.00,130772.00,125739.00,130772.00,108720.00,130772.00,130772.00,130772.00,130737.00,130772.00,130772.00,130772.00,130772.00,130772.00,130767.00,129032.00,128980.00,130772.00,130772.00,125739.00,130772.00,130772.00,130772.00,130772.00,129136.00,26218.00,115178.00,38001.00,130772.00,130772.00,130772.00,130772.00,130772.00,130772.00,130772.00,130772.00,130772.00,130772.00,127989.0,130772.0,130772.00,130772.00,130772.00,129031.00,130772.00,130772.00,130772.00,130772.00,130772.00,130772.00,18087.00,18087.00,18087.00,18087.00,17768.00,18087.00,18087.00,18087.00,18087.00,6184.00,44.0,44.00,44.0,44.00,41.00,44.00,44.00,71.00,71.00,71.00
mean,NaN,15908.83,15908.83,15905.58,463.11,78665.13,NaN,NaN,19.29,0.24,0.45,36.63,82.72,11.44,0.13,15531.62,22.47,12124.26,12121.90,4709.79,4708.74,3555.90,1150.34,0.50,3.06,0.51,1536.45,0.02,46.08,1.0,126482.26,19.35,0.0,196.19,140239.86,0.89,2.67,0.68,1.52,21.54,34879.27,68.30,1.28,2.73,5575.08,54.07,36963.61,1.11,1.46,2.03,4.53,13461.57,14467.05,49.56,0.01,1.57,122.30,172.22,14.85,8.74,1.31,25.18,40.37,7.42,37.55,0.48,3.55,5.30,4.76,7.01,8.12,8.14,12.85,5.27,11.42,0.0,0.0,0.06,2.09,94.43,32.67,0.12,0.01,180023.71,50716.80,25425.28,44900.21,33428.53,0.64,1.46,11.54,57.56,3.04,12.49,0.04,0.08,36.52,3.0,175.66,3.0,7.32,545.30,16417.42,172.33,7789.36,59.94,18.79
std,NaN,10129.85,10129.85,10128.95,286.20,87535.10,NaN,NaN,19.48,0.75,0.72,21.69,23.12,5.98,0.37,22713.90,12.16,9320.34,9319.58,5138.45,5137.60,4953.85,969.01,5.39,145.62,25.60,4557.13,0.15,21.30,0.0,97892.66,8.13,0.0,1599.13,162249.81,1.12,2.95,0.94,1.57,26.24,44702.37,23.79,1.54,2.65,6431.32,21.32,37848.08,1.55,2.65,2.45,3.31,16936.07,19149.87,29.28,0.09,192.47,55.47,100.10,18.50,9.60,1.68,33.47,22.16,6.05,21.88,1.39,2.40,3.43,3.13,4.49,7.40,4.96,7.89,3.38,5.98,0.0,0.0,0.42,1.94,9.27,35.12,0.34,0.15,182545.83,52472.96,25327.50,47136.83,28331.62,1.00,1.73,6.74,25.67,3.33,8.19,0.42,0.

In [9]:
display(df["annual_inc"].nlargest(5))
display((df["annual_inc"] > 1000000).sum())

116065    9930475.0
28140     9300000.0
128519    9000000.0
28811     7600033.0
8839      6200000.0
Name: annual_inc, dtype: float64

np.int64(67)

#### Observation

`annual_inc` has a very long upper tail. There are 67 borrowers with reported annual income above $1M, and the largest values are several million dollars. These may be valid high-income borrowers, but they should be inspected or capped later because they can strongly affect averages and model behavior.

In [10]:
df.describe(include="object")

C:\Users\janis\AppData\Local\Temp\ipykernel_16540\702825166.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include="object")


,id,term,int_rate,grade,sub_grade,emp_title,emp_length,home_ownership,verification_status,issue_d,loan_status,pymnt_plan,purpose,title,zip_code,addr_state,earliest_cr_line,revol_util,initial_list_status,last_pymnt_d,next_pymnt_d,last_credit_pull_d,application_type,verification_status_joint,sec_app_earliest_cr_line,hardship_flag,hardship_type,hardship_reason,hardship_status,hardship_start_date,hardship_end_date,payment_plan_start_date,hardship_loan_status,disbursement_method,debt_settlement_flag,debt_settlement_flag_date,settlement_status,settlement_date
count,2,130772,130772,130772,130772,120110,120304,130772,130772,130772,130772,130772,130772,130772,130772,130772,130772,130614,130772,130586,117403,130772,130772,17417,18087,130772,44,44,44,44,44,44,44,130772,130772,71,71,71
unique,2,2,57,7,35,44609,11,4,3,3,7,2,12,12,874,50,640,1063,2,10,4,11,2,3,562,2,1,6,3,5,6,6,4,2,2,6,3,6
top,Total amount funded in policy code 1: 2080429200,36 months,5.31%,B,B4,Teacher,10+ years,MORTGAGE,Source Verified,May-2018,Current,n,debt_consolidation,Debt consolidation,112xx,CA,Aug-2006,0%,w,Jan-2019,Feb-2019,Jan-2019,Individual,Not Verified,Aug-2006,N,INTEREST ONLY-3 MONTHS DEFERRAL,NATURAL_DISASTER,COMPLETED,Oct-2018,Jan-2019,Oct-2018,Current,Cash,N,Jan-2019,ACTIVE,Jan-2019
freq,1,91506,6236,38246,8575,2374,43815,61720,53100,46311,113004,130770,64141,64141,1450,17417,1118,1463,113285,114446,117362,122205,112685,7796,158,130758,44,29,27,20,17,16,26,116353,130701,40,63,26


### Investigating Zero-Income Loans

Some loans have `annual_inc == 0`. This is worth checking separately because zero income may represent a data-quality issue or unusual borrower reporting, not simply a normal low-income segment.

In [11]:
(df["annual_inc"] == 0).sum()

np.int64(322)

In [12]:
zero_income = df[df["annual_inc"] == 0]
zero_income[
[
  "annual_inc",
  "loan_amnt",
  "grade",
  "home_ownership",
  "emp_length",
  "verification_status",
  "loan_status"
]]

,annual_inc,loan_amnt,grade,home_ownership,emp_length,verification_status,loan_status
72,0.0,24000.0,D,MORTGAGE,10+ years,Verified,Current
197,0.0,16500.0,A,MORTGAGE,NaN,Not Verified,Current
1103,0.0,20000.0,A,MORTGAGE,NaN,Not Verified,Current
2342,0.0,10000.0,B,RENT,NaN,Verified,Current
2502,0.0,11000.0,B,MORTGAGE,NaN,Not Verified,Charged Off
...,...,...,...,...,...,...,...
128786,0.0,20000.0,B,RENT,NaN,Not Verified,Fully Paid
128792,0.0,4675.0,B,RENT,NaN,Verified,Current
128793,0.0,35000.0,C,MORTGAGE,NaN,Not Verified,Current
129710,0.0,2500.0,B,MORTGAGE,NaN,Not Verified,Current


In [13]:
display(zero_income.groupby("verification_status")
.size()
.div(len(zero_income))
.mul(100)
.round(2)
)

#compare loan amt
display(zero_income["loan_amnt"].describe().round(2))

min_income = df[df["annual_inc"].between(1, 24999) ]
display(min_income["loan_amnt"].describe().round(2))

verification_status
Not Verified       86.34
Source Verified     2.48
Verified           11.18
dtype: float64

count      322.00
mean     20553.57
std      10597.01
min       2000.00
25%      12000.00
50%      20000.00
75%      30000.00
max      40000.00
Name: loan_amnt, dtype: float64

count     5456.00
mean      9852.14
std       7975.71
min       1000.00
25%       4800.00
50%       7200.00
75%      12000.00
max      40000.00
Name: loan_amnt, dtype: float64

In [14]:

#compare loan status with low income group
display(
  zero_income.groupby("loan_status")
  .size()
  .div(len(zero_income))
  .mul(100)
  .round(2)
  .to_frame("percentage")
)

display(
  min_income.groupby("loan_status")
  .size()
  .div(len(min_income))
  .mul(100)
  .round(2)
  .to_frame("percentage")
)



,percentage
loan_status,
Charged Off,1.55
Current,88.20
Fully Paid,6.21
In Grace Period,1.24
Late (16-30 days),0.62
Late (31-120 days),2.17


,percentage
loan_status,
Charged Off,2.07
Current,86.82
Fully Paid,7.22
In Grace Period,0.99
Late (16-30 days),0.57
Late (31-120 days),2.33


### Observation

There are 322 zero-income loans. Their loan amounts are surprisingly high compared with borrowers earning $1-$24,999: the zero-income median loan amount is $20,000 versus $7,200 for the low-income comparison group. Most zero-income rows are `Not Verified`, which supports treating this group cautiously as a possible reporting or data-quality issue.

The loan-status mix for zero-income borrowers is similar to the low-income comparison group, but the zero-income group is small and mostly still `Current`, so this does not prove the records are reliable.

### DTI Exploration


In [15]:
# Basic DTI distribution
df["dti"].describe(percentiles=[0.01, 0.05, 0.25, 0.5, 0.75, 0.95, 0.99]).round(2)

count    130446.00
mean         19.29
std          19.48
min           0.00
1%            1.05
5%            3.84
25%          11.09
50%          17.38
75%          24.73
95%          36.81
99%          59.85
max         999.00
Name: dti, dtype: float64

In [16]:
dti_summary = pd.Series({
  "missing": df["dti"].isna().sum(),
  "missing_pct": (df["dti"].isna().mean() * 100).round(2),
  "zero": (df["dti"] == 0).sum(),
  "over_100": (df["dti"] > 100).sum(),
  "equal_999": (df["dti"] == 999).sum()
})
dti_summary

missing        328.00
missing_pct      0.25
zero           260.00
over_100       394.00
equal_999       13.00
dtype: float64

In [17]:
cols = [
    "dti",
    "annual_inc",
    "loan_amnt",
    "installment",
    "verification_status",
    "home_ownership",
    "purpose",
    "grade",
    "sub_grade",
    "loan_status"
    
]

df.loc[
  df["dti"] > 100,
  cols
].sort_values(by = "dti", ascending=False)

,dti,annual_inc,loan_amnt,installment,verification_status,home_ownership,purpose,grade,sub_grade,loan_status
122321,999.00,400.0,20000.0,433.86,Verified,MORTGAGE,home_improvement,B,B4,Current
10163,999.00,20.0,20000.0,641.41,Verified,RENT,credit_card,B,B1,Current
82929,999.00,1000.0,15000.0,325.39,Source Verified,MORTGAGE,credit_card,B,B4,Fully Paid
80316,999.00,100.0,30500.0,997.09,Not Verified,OWN,credit_card,B,B4,Current
59374,999.00,700.0,20000.0,470.78,Not Verified,MORTGAGE,debt_consolidation,C,C3,Current
...,...,...,...,...,...,...,...,...,...,...
42817,100.90,12000.0,25000.0,614.08,Not Verified,MORTGAGE,vacation,C,C5,Current
93802,100.47,29300.0,13500.0,314.62,Verified,MORTGAGE,home_improvement,C,C3,Fully Paid
89609,100.33,40000.0,40000.0,798.49,Verified,MORTGAGE,credit_card,A,A4,Current
128556,100.30,12000.0,16000.0,514.62,Source Verified,MORTGAGE,debt_consolidation,F,F2,Late (16-30 days)


#### Observation

DTI missingness is low: 328 rows, about 0.25% of the dataset. The bigger issue is extreme values. The 99th percentile is 59.85, but the maximum is 999, with 394 rows above 100 and 13 rows exactly equal to 999. These values should be treated as extreme or special cases during preprocessing, for example by flagging high DTI and possibly capping the numeric value.

#### Non numerical column inspection

In [18]:
cat_cols = df.select_dtypes(include=["object", "string"]).columns

cat_summary = pd.DataFrame({
  "dtype": df[cat_cols].dtypes,
  "missing": df[cat_cols].isna().sum(),
  "missing_pct": df[cat_cols].isna().mean() * 100,
  "unique": df[cat_cols].nunique(dropna=False)
}).sort_values("unique", ascending=False)

cat_summary

,dtype,missing,missing_pct,unique
emp_title,str,10664,8.154526,44610
revol_util,str,160,0.122348,1064
zip_code,str,2,0.001529,875
earliest_cr_line,str,2,0.001529,641
sec_app_earliest_cr_line,str,112687,86.169269,563
int_rate,str,2,0.001529,58
addr_state,str,2,0.001529,51
sub_grade,str,2,0.001529,36
title,str,2,0.001529,13
purpose,str,2,0.001529,13


#### Observations

- 4 variables were identified as numeric values stored as strings.
- 5 variables can be converted to datetime format.
- Most categorical variables exhibit low cardinality (<50 categories) and are suitable for encoding.
- emp_title (44,610 unique values) has a very high-cardinality and will be evaluated separately.
- Several hardship and settlement related variables contain >99% missing values and appear to contain post-origination information. These variables are potential sources of target leakage and will likely be excluded from modeling.